jerry-core/mem/mem-poolman.cpp

```diff
@@ -193,88 +194,128 @@ mem_pools_finalize (void)
 
+#ifdef JERRY_VALGRIND
   /*
-   * At first pass collect pointers to those of free chunks that are first at their pools
-   * to separate lists (collection-time pool lists) and change them to headers of corresponding pools
+   * If the chunk is not free, there may be undefined bytes at hint_magic_num and traversal_check_flag fields.
+   *
+   * Although, it is correct for the routine, valgrind issues warning about using uninitialized data
+   * in conditional expression. To suppress the false-positive warning, the chunk is temporarily marked
+   * as defined, and after reading hint magic number and list identifier, valgrind state of the chunk is restored.
    */
+  uint8_t vbits[MEM_POOL_CHUNK_SIZE];
+  unsigned status;
+
+  status = VALGRIND_GET_VBITS (pool_first_chunk_p, vbits, MEM_POOL_CHUNK_SIZE);
+  JERRY_ASSERT (status == 0 || status == 1);
+
+  VALGRIND_DEFINED_SPACE (pool_first_chunk_p, MEM_POOL_CHUNK_SIZE);
+#endif /* JERRY_VALGRIND */
+
+  uint16_t magic_num_field = pool_first_chunk_p->u.pool_gc.hint_magic_num;
+  bool traversal_check_flag = pool_first_chunk_p->u.pool_gc.traversal_check_flag;
 
+#ifdef JERRY_VALGRIND
+  status = VALGRIND_SET_VBITS (pool_first_chunk_p, vbits, MEM_POOL_CHUNK_SIZE);
+  JERRY_ASSERT (status == 0 || status == 1);
+#endif /* JERRY_VALGRIND */
+
+  *out_magic_num_field_value_p = magic_num_field;
+  *out_traversal_check_flag_p = traversal_check_flag;
+} /* mem_pools_collect_read_magic_num_and_flag */
+
+/**
+ * Collect chunks from empty pools and free the pools
+ */
+void
+mem_pools_collect_empty (void)
+{
   /*
-   * Number of collection-time pool lists
+   * Hint magic number in header of pools with free pool-first chunks
    */
-  constexpr uint32_t pool_lists_number = 8;
+  const uint16_t hint_magic_num_value = 0x7e89;
 
   /*
-   * Collection-time pool lists
+   * Collection-time chunk lists
    */
-  mem_pool_chunk_t *pool_lists_p[pool_lists_number];
-  for (uint32_t i = 0; i < pool_lists_number; i++)
-  {
-    pool_lists_p[i] = NULL;
-  }
+  mem_pool_chunk_t *first_chunks_list_p = NULL;
+  mem_pool_chunk_t *non_first_chunks_list_p = NULL;
 
   /*
-   * Number of the pools, included into the lists
+   * At first stage collect free pool-first chunks to separate collection-time lists
+   * and change their layout from mem_pool_chunk_t::u::free to mem_pool_chunk_t::u::pool_gc
    */
-  uint32_t pools_in_lists_number = 0;
-
-  for (mem_pool_chunk_t *free_chunk_iter_p = mem_free_chunk_p, *prev_free_chunk_p = NULL, *next_free_chunk_p;
-       free_chunk_iter_p != NULL;
-       free_chunk_iter_p = next_free_chunk_p)
   {
-    mem_pool_chunk_t *pool_start_p = (mem_pool_chunk_t *) mem_heap_get_chunked_block_start (free_chunk_iter_p);
+    mem_pool_chunk_t tmp_header;
+    tmp_header.u.free.next_p = mem_free_chunk_p;
+
+    for (mem_pool_chunk_t *free_chunk_iter_p = tmp_header.u.free.next_p,
+                          *prev_free_chunk_p = &tmp_header,
+                          *next_free_chunk_p;
```

Comment:
Minor thing: the indentation of the `for` is different in the modified parts troughout the patch